### PR TITLE
Add GetTool and ListTools API

### PR DESCRIPTION
**Pull request by jackiehanyang**
### Description
This pr merges GetTool and ListTools API
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jackiehanyang** [comment]
CI failed due to flaky integration tests issue. 

On github CI, these two IT failed with message:

```
org.opensearch.ml.rest.RestMLRemoteInferenceIT > testCohereGenerateTextModel FAILED
    org.opensearch.client.ResponseException: method [POST], host [http://[::1]:58190], URI [/_plugins/_ml/models/yAnva4wBbaab9Fw_3aDZ/_predict], status line [HTTP/1.1 429 Too Many Requests]
    {"error":{"root_cause":[{"type":"status_exception","reason":"Error from remote service: {\"message\":\"You are using a Trial key, which is limited to 1000 API calls / month. You can continue to use the Trial key for free or upgrade to a Production key with higher rate limits at 'https://dashboard.cohere.ai/api-keys'. Contact us on 'https://discord.gg/XW44jPfYJu' or email us at support@cohere.com with any questions\"}"}],"type":"status_exception","reason":"Error from remote service: {\"message\":\"You are using a Trial key, which is limited to 1000 API calls / month. You can continue to use the Trial key for free or upgrade to a Production key with higher rate limits at 'https://dashboard.cohere.ai/api-keys'. Contact us on 'https://discord.gg/XW44jPfYJu' or email us at support@cohere.com with any questions\"}"},"status":429}
        at __randomizedtesting.SeedInfo.seed([AAFF27FFAF877CE3:1381DDEC956996A0]:0)
        at app//org.opensearch.client.RestClient.convertResponse(RestClient.java:385)
        at app//org.opensearch.client.RestClient.performRequest(RestClient.java:355)
        at app//org.opensearch.client.RestClient.performRequest(RestClient.java:330)
        at app//org.opensearch.ml.utils.TestHelper.makeRequest(TestHelper.java:167)
        at app//org.opensearch.ml.utils.TestHelper.makeRequest(TestHelper.java:140)
        at app//org.opensearch.ml.utils.TestHelper.makeRequest(TestHelper.java:129)
        at app//org.opensearch.ml.rest.RestMLRemoteInferenceIT.predictRemoteModel(RestMLRemoteInferenceIT.java:692)
        at app//org.opensearch.ml.rest.RestMLRemoteInferenceIT.testCohereGenerateTextModel(RestMLRemoteInferenceIT.java:526)
REPRODUCE WITH: gradlew ':opensearch-ml-plugin:integTest' --tests "org.opensearch.ml.rest.RestMLRemoteInferenceIT.testCohereGenerateTextModel" -Dtests.seed=AAFF27FFAF877CE3 -Dtests.security.manager=false -Dtests.locale=es-US -Dtests.timezone=America/Cancun -Druntime.java=11

org.opensearch.ml.rest.RestMLRemoteInferenceIT > testCohereClassifyModel FAILED
    org.opensearch.client.ResponseException: method [POST], host [http://127.0.0.1:5[818](https://github.com/opensearch-project/ml-commons/actions/runs/7218135230/job/19667071325?pr=1764#step:5:819)9], URI [/_plugins/_ml/models/zQnwa4wBbaab9Fw_L6C1/_predict], status line [HTTP/1.1 429 Too Many Requests]
    {"error":{"root_cause":[{"type":"status_exception","reason":"Error from remote service: {\"message\":\"You are using a Trial key, which is limited to 1000 API calls / month. You can continue to use the Trial key for free or upgrade to a Production key with higher rate limits at 'https://dashboard.cohere.ai/api-keys'. Contact us on 'https://discord.gg/XW44jPfYJu' or email us at support@cohere.com with any questions\"}"}],"type":"status_exception","reason":"Error from remote service: {\"message\":\"You are using a Trial key, which is limited to 1000 API calls / month. You can continue to use the Trial key for free or upgrade to a Production key with higher rate limits at 'https://dashboard.cohere.ai/api-keys'. Contact us on 'https://discord.gg/XW44jPfYJu' or email us at support@cohere.com with any questions\"}"},"status":429}
        at __randomizedtesting.SeedInfo.seed([AAFF27FFAF877CE3:8CFDA771A06EE7E5]:0)
        at app//org.opensearch.client.RestClient.convertResponse(RestClient.java:385)
        at app//org.opensearch.client.RestClient.performRequest(RestClient.java:355)
        at app//org.opensearch.client.RestClient.performRequest(RestClient.java:330)
        at app//org.opensearch.ml.utils.TestHelper.makeRequest(TestHelper.java:167)
        at app//org.opensearch.ml.utils.TestHelper.makeRequest(TestHelper.java:140)
        at app//org.opensearch.ml.utils.TestHelper.makeRequest(TestHelper.java:129)
        at app//org.opensearch.ml.rest.RestMLRemoteInferenceIT.predictRemoteModel(RestMLRemoteInferenceIT.java:692)
        at app//org.opensearch.ml.rest.RestMLRemoteInferenceIT.testCohereClassifyModel(RestMLRemoteInferenceIT.java:636)
REPRODUCE WITH: gradlew ':opensearch-ml-plugin:integTest' --tests "org.opensearch.ml.rest.RestMLRemoteInferenceIT.testCohereClassifyModel" -Dtests.seed=AAFF27FFAF877CE3 -Dtests.security.manager=false -Dtests.locale=es-US -Dtests.timezone=America/Cancun -Druntime.java=11
```

on my cloud dev desktop, this IT failed ` - org.opensearch.ml.action.prediction.PredictionITTests.testPredictionWithoutDataset_KMeans`

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java [hunk 1]
```diff
@@ -490,6 +506,36 @@ public Collection<Object> createComponents(
             encryptor
         );
 
+        toolFactories = new HashMap<>();
+
+        // MLModelTool.Factory.getInstance().init(client);
+        // MathTool.Factory.getInstance().init(scriptService);
+        // VectorDBTool.Factory.getInstance().init(client, xContentRegistry);
+        // NeuralSparseTool.Factory.getInstance().init(client, xContentRegistry);
+        // AgentTool.Factory.getInstance().init(client);
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java:515]: ?

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java [hunk 2]
```diff
@@ -490,6 +506,36 @@ public Collection<Object> createComponents(
...
+        // VisualizationsTool.Factory.getInstance().init(client);
+        // RAGTool.Factory.getInstance().init(client, xContentRegistry);
+        // SearchAlertsTool.Factory.getInstance().init(client);
+        // IndexMappingTool.Factory.getInstance().init(client);
+
+        // toolFactories.put(MLModelTool.TYPE, MLModelTool.Factory.getInstance());
+        // toolFactories.put(MathTool.TYPE, MathTool.Factory.getInstance());
+        // toolFactories.put(VectorDBTool.TYPE, VectorDBTool.Factory.getInstance());
+        // toolFactories.put(NeuralSparseTool.TYPE, NeuralSparseTool.Factory.getInstance());
+        // toolFactories.put(AgentTool.TYPE, AgentTool.Factory.getInstance());
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java:527]: ?

File: plugin/src/main/java/org/opensearch/ml/tools/GetToolTransportAction.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+     * @param task
+     * @param request
+     * @param listener
+     */
+    @Override
+    protected void doExecute(Task task, ActionRequest request, ActionListener<MLToolGetResponse> listener) {
+        MLToolGetRequest mlToolGetRequest = MLToolGetRequest.fromActionRequest(request);
+        String toolName = mlToolGetRequest.getToolName();
+        List<ToolMetadata> externalTools = mlToolGetRequest.getExternalTools();
+        List<ToolMetadata> toolsList = new ArrayList<>(
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/tools/GetToolTransportAction.java:45]: Are these default tool list which we will have in every agents? What's the rationale behind this?

File: common/src/test/java/org/opensearch/ml/common/transport/tools/MLToolGetResponseTests.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+//    @Test
+//    public void toXContentTest() throws IOException {
+//        MLToolGetResponse mlToolGetResponse = MLToolGetResponse.builder().toolMetadata(toolMetadata).build();
+//        XContentBuilder builder = XContentFactory.contentBuilder(XContentType.JSON);
+//        mlToolGetResponse.toXContent(builder, ToXContent.EMPTY_PARAMS);
+//        assertNotNull(builder);
+//        String jsonStr = Strings.toString(builder);
+//        assertEquals(jsonStr, "{\"name\":\"MathTool\",\"description\":\"Use this tool to calculate any math problem.\"}");
+//    }
+}
```
Comment (Hailong-am) [inline common/src/test/java/org/opensearch/ml/common/transport/tools/MLToolGetResponseTests.java:48]: why it's commented out? if it's not needed, we can remove it.

File: common/src/test/java/org/opensearch/ml/common/transport/tools/MLToolsListResponseTests.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+//    @Test
+//    public void toXContentTest() throws IOException {
+//        MLToolsListResponse mlToolsListResponse = MLToolsListResponse.builder().toolMetadata(toolMetadataList).build();
+//        XContentBuilder builder = XContentFactory.contentBuilder(XContentType.JSON);
+//        mlToolsListResponse.toXContent(builder, ToXContent.EMPTY_PARAMS);
+//        assertNotNull(builder);
+//        String jsonStr = Strings.toString(builder);
+//        assertEquals("{\"SearchWikipediaTool\":\"Useful when you need to use this tool to search general knowledge on wikipedia.\"}", jsonStr);
+//    }
+}
```
Comment (Hailong-am) [inline common/src/test/java/org/opensearch/ml/common/transport/tools/MLToolsListResponseTests.java:55]: same here

File: common/src/main/java/org/opensearch/ml/common/ToolMetadata.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+public class ToolMetadata implements ToXContentObject, Writeable {
+
+    public static final String TOOL_NAME_FIELD = "name";
+    public static final String TOOL_DESCRIPTION_FIELD = "description";
+
+    @Getter
+    private String name;
+    @Getter
+    private String description;
```
Comment (Hailong-am) [inline common/src/main/java/org/opensearch/ml/common/ToolMetadata.java:26]: do we need tool type here?

File: plugin/src/main/java/org/opensearch/ml/rest/RestMLGetToolAction.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+     * @throws IOException if an I/O exception occurred parsing the request and preparing for
+     *                     execution
+     */
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        List<ToolMetadata> toolList = new ArrayList<>();
+        toolFactories
+            .forEach((key, value) -> toolList.add(ToolMetadata.builder().name(key).description(value.getDefaultDescription()).build()));
+        String toolName = getParameterId(request, PARAMETER_TOOL_NAME);
+        MLToolGetRequest mlToolGetRequest = MLToolGetRequest.builder().toolName(toolName).externalTools(toolList).build();
```
Comment (Hailong-am) [inline plugin/src/main/java/org/opensearch/ml/rest/RestMLGetToolAction.java:67]: should we place external tool logic in transport layer?

File: plugin/src/main/java/org/opensearch/ml/rest/RestMLListToolsAction.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+     * @return the action to execute
+     * @throws IOException if an I/O exception occurred parsing the request and preparing for
+     *                     execution
+     */
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        List<ToolMetadata> toolList = new ArrayList<>();
+        toolFactories
+            .forEach((key, value) -> toolList.add(ToolMetadata.builder().name(key).description(value.getDefaultDescription()).build()));
+        MLToolsListRequest mlToolsGetRequest = MLToolsListRequest.builder().externalTools(toolList).build();
```
Comment (Hailong-am) [inline plugin/src/main/java/org/opensearch/ml/rest/RestMLListToolsAction.java:64]: should we place external tool logic in transport layer?

File: plugin/src/main/java/org/opensearch/ml/tools/GetToolTransportAction.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+                        .description("Useful when you need to use this tool to search general knowledge on wikipedia.")
+                        .build()
+                )
+        );
+        toolsList.addAll(externalTools);
+        ToolMetadata theTool = toolsList
+            .stream()
+            .filter(tool -> tool.getName().equals(toolName))
+            .findFirst()
+            .orElseThrow(NoSuchElementException::new);
```
Comment (Hailong-am) [inline plugin/src/main/java/org/opensearch/ml/tools/GetToolTransportAction.java:69]: should this line included in `try` block?

File: plugin/src/main/java/org/opensearch/ml/tools/ListToolsTransportAction.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+                    ToolMetadata
+                        .builder()
+                        .name("SearchIndexTool")
+                        .description(
+                            "Useful for when you don't know answer for some question or need to search my private data in OpenSearch index."
+                        )
+                        .build(),
+                    ToolMetadata
+                        .builder()
+                        .name("SearchWikipediaTool")
```
Comment (Hailong-am) [inline plugin/src/main/java/org/opensearch/ml/tools/ListToolsTransportAction.java:55]: do we have this tool?

Comment (jackiehanyang) [reply plugin/src/main/java/org/opensearch/ml/tools/ListToolsTransportAction.java:55]: @ylwu-amzn Back in August, when this project was still experimental, we hardcoded four internal tools along with one external tool. This setup is still present in the feature branch, so I've included them in this PR. Should we modify this part now, or should we proceed with the merge from the feature branch as it is? Personally, I prefer handling one task at a time - avoiding refactoring during the merge process.

Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/tools/ListToolsTransportAction.java:55]: Let's modify the code now. I know we merged a lot of code to feature branch for fast iteration. When merge to main branch, let's hold the bar and modify the code.

File: plugin/src/main/java/org/opensearch/ml/rest/RestMLGetToolAction.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+     * @return the action to execute
+     * @throws IOException if an I/O exception occurred parsing the request and preparing for
+     *                     execution
+     */
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        List<ToolMetadata> toolList = new ArrayList<>();
+        toolFactories
+            .forEach((key, value) -> toolList.add(ToolMetadata.builder().name(key).description(value.getDefaultDescription()).build()));
+        String toolName = getParameterId(request, PARAMETER_TOOL_NAME);
```
Comment (Zhangxunmt) [inline plugin/src/main/java/org/opensearch/ml/rest/RestMLGetToolAction.java:66]: Is it necessary to validate the tooName here, if only a set of names are allowed?

File: plugin/src/main/java/org/opensearch/ml/tools/GetToolTransportAction.java (hunk 9 above)
Comment (Zhangxunmt) [reply plugin/src/main/java/org/opensearch/ml/tools/GetToolTransportAction.java:69]: +1. Why not include the lines from 44 to 69 into the try block? At least the NoSuchElementException exception can be catch then.

File: plugin/src/main/java/org/opensearch/ml/rest/RestMLGetToolAction.java (hunk 7 above)
Comment (jackiehanyang) [reply plugin/src/main/java/org/opensearch/ml/rest/RestMLGetToolAction.java:67]: need this logic here to prepare the request

File: plugin/src/main/java/org/opensearch/ml/tools/ListToolsTransportAction.java (hunk 10 above)
Comment (jackiehanyang) [reply plugin/src/main/java/org/opensearch/ml/tools/ListToolsTransportAction.java:55]: leaving refactoring related comments as unresolved and created this issue to track it - https://github.com/opensearch-project/ml-commons/issues/1802

File: common/src/main/java/org/opensearch/ml/common/ToolMetadata.java (hunk 6 above)
Comment (jackiehanyang) [reply common/src/main/java/org/opensearch/ml/common/ToolMetadata.java:26]: Please leave a comment in this [issue](https://github.com/opensearch-project/ml-commons/issues/1802) if you believe we need to add tool type here

File: plugin/src/main/java/org/opensearch/ml/rest/RestMLGetToolAction.java (hunk 7 above)
Comment (Hailong-am) [reply plugin/src/main/java/org/opensearch/ml/rest/RestMLGetToolAction.java:67]: make sense for rest API, but we have this in ml-client also, that's transport layer.https://github.com/opensearch-project/ml-commons/blob/feature/agent_framework_dev/client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java#L268C5-L279, i didn't see externalTools passed in, if i use ml-client to get a external tool, will it return correct result?

File: common/src/main/java/org/opensearch/ml/common/ToolMetadata.java (hunk 6 above)
Comment (Hailong-am) [reply common/src/main/java/org/opensearch/ml/common/ToolMetadata.java:26]: Not sure who will be the consumer of this API, tool name is not enough for register a agent, if it should used by ai flow.

File: client/src/test/java/org/opensearch/ml/client/MachineLearningClientTest.java [hunk 12]
```diff
@@ -173,6 +176,16 @@ public void execute(FunctionName name, Input input, ActionListener<MLExecuteTask
                 listener.onResponse(mlExecuteTaskResponse);
             }
 
+            @Override
+            public void listTools(ActionListener<List<ToolMetadata>> listener) {
+                listener.onResponse(null);
```
Comment (dhrubo-os) [inline client/src/test/java/org/opensearch/ml/client/MachineLearningClientTest.java:181]: why null?

File: common/src/test/java/org/opensearch/ml/common/transport/tools/MLToolsListResponseTests.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        BytesStreamOutput bytesStreamOutput = new BytesStreamOutput();
+        MLToolsListResponse response = MLToolsListResponse.builder().toolMetadata(toolMetadataList).build();
+        response.writeTo(bytesStreamOutput);
+        MLToolsListResponse parsedResponse = new MLToolsListResponse(bytesStreamOutput.bytes().streamInput());
+        assertNotEquals(response.toolMetadataList, parsedResponse.toolMetadataList);
+        assertEquals(response.toolMetadataList.get(0).getName(), parsedResponse.toolMetadataList.get(0).getName());
+        assertEquals(response.toolMetadataList.get(0).getDescription(), parsedResponse.toolMetadataList.get(0).getDescription());
+    }
+
+//    @Test
```
Comment (dhrubo-os) [inline common/src/test/java/org/opensearch/ml/common/transport/tools/MLToolsListResponseTests.java:46]: remove?

File: plugin/src/main/java/org/opensearch/ml/tools/ListToolsTransportAction.java (hunk 10 above)
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/tools/ListToolsTransportAction.java:55]: In that case, let's work on the created issue before merging this code.

**Pull request closed without merging**